scripts/unix/download-tool.sh

```diff
@@ -0,0 +1,82 @@
+#!/bin/bash
+
+###
+# Copyright 2016 resin.io
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###
+
+set -u
+set -e
+set -x
+
+function check_dep() {
+  if ! command -v $1 2>/dev/null 1>&2; then
+    echo "Dependency missing: $1" 1>&2
+    exit 1
+  fi
+}
+
+check_dep wget
+check_dep md5sum
+
+function usage() {
+  echo "Usage: $0"
+  echo ""
+  echo "Options"
+  echo ""
+  echo "    -u <url>"
+  echo "    -m <md5 hash>"
+  echo "    -o <output>"
+  exit 1
+}
+
+ARGV_URL=""
+ARGV_MD5_HASH=""
+ARGV_OUTPUT=""
+
+while getopts ":u:m:o:" option; do
+  case $option in
+    u) ARGV_URL="$OPTARG" ;;
+    m) ARGV_MD5_HASH="$OPTARG" ;;
+    o) ARGV_OUTPUT="$OPTARG" ;;
+    *) usage ;;
+  esac
+done
+
+if [ -z "$ARGV_URL" ] || \
+   [ -z "$ARGV_MD5_HASH" ] || \
+   [ -z "$ARGV_OUTPUT" ]
+then
+  usage
+fi
+
+function md5sum_matches() {
+  local file=$1
+  local hash=$2
+	test "$(md5sum $file | cut -d ' ' -f1)" = "$hash"
+}
+
+if [ -e "$ARGV_OUTPUT" ] && md5sum_matches "$ARGV_OUTPUT" "$ARGV_MD5_HASH"; then
+  echo "Re-using from cache"
+  exit 0
+fi
+
+wget --no-check-certificate "$ARGV_URL" -O "$ARGV_OUTPUT"
+chmod +x "$ARGV_OUTPUT"
```

Comment:
IMHO this `chmod` doesn't belong in this script (since it may well be downloading things that aren't executable), but in the scripts that call this script.